### ISSUE TITLE
SequenceEqual broken for concurrent use?

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#47016]
Repro:
```C#
using System;
using System.Security.Cryptography;
using System.Threading;

class Program
{
    static void Main()
    {
        static void Work(object _)
        {
            while (true)
            {
                Span<byte> buffer1 = RandomNumberGenerator.GetBytes(1024 * 1024);
                Span<byte> buffer2 = new byte[buffer1.Length];
                buffer1.CopyTo(buffer2);

                if (!buffer1.SequenceEqual(buffer2)) // should never be false...
                {
                    while (!buffer1.SequenceEqual(buffer2)) ; // but if it is, it certainly shouldn't go back to true...
                    throw new Exception("What in the world just happened?!");
                }
            }
        }

        ThreadPool.QueueUserWorkItem(Work);
        ThreadPool.QueueUserWorkItem(Work);
        Console.ReadLine();
    }
}
```

This runs without error on .NET 5.

On master on my Windows machine (I've not tried other OSes), within a few seconds I get a failure like this:
```
Unhandled exception. System.Exception: SequenceEquals changed from false to true?!
   at Program.Work() in C:\Users\stoub\source\repos\ConsoleApp17\ConsoleApp17\Program.cs:line 25
   at Program.<>c.<Main>b__0_0(Object _) in C:\Users\stoub\source\repos\ConsoleApp17\ConsoleApp17\Program.cs:line 10
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\ThreadPool.cs:line 977
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\ThreadPool.cs:line 705
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\PortableThreadPool.WorkerThread.cs:line 56
   at System.Threading.Thread.StartCallback() in D:\repos\runtime\src\coreclr\System.Private.CoreLib\src\System\Threading\Thread.CoreCLR.cs:line 105
```

Note that if I remove one of the ThreadPool calls so that there's no concurrency, it's never failed for me; only once I introduce the parallelism does it start to fail.

JIT bug?  GC? Some kind of latent issue in SequenceEquals?

cc: @geoffkizer, @jkotas

**Comment by stephentoub** [comment]
As a follow-up, this still fails if I pin both buffers around the whole copy+compare operation.  And it still fails if I move one of the allocations out of the loop.  But it stops reproing for me if I move both allocations out of the loop.  So maybe something to do with alignment?

**Comment by jkotas** [comment]
This looks like a thread suspension bug to me.  @VSadov Could you please take a look since you have done refactoring of that code recently?

**Comment by VSadov** [comment]
Will take a look.

**Comment by VSadov** [comment]
Hmm, does not compile for me.
> RandomNumberGenerator.GetBytes(1024 * 1024);

Is that a new API in 6.0 ?

**Comment by jkotas** [comment]
Here is a simpler repro that does not use that API:
```csharp
using System;
using System.Threading;

class Program
{
    static object o = null;

    static void Main()
    {
        static void Work(object _)
        {
            while (true)
            {
                Span<byte> buffer1 = new byte[1024 * 1024];
                for (int i = 0; i < buffer1.Length; i++) buffer1[i] = (byte)i;
                Span<byte> buffer2 = new byte[buffer1.Length];
                buffer1.CopyTo(buffer2);

                if (!buffer1.SequenceEqual(buffer2)) // should never be false...
                {
                    o.ToString();
                    throw new Exception("What in the world just happened?!");
                }
            }
        }

        ThreadPool.QueueUserWorkItem(Work);
        ThreadPool.QueueUserWorkItem(Work);
        Thread.Sleep(-1);
    }
}
```

**Comment by VSadov** [comment]
That builds, but so far I do not see failures. Tried more tasks, and server GC as well. 
(I am using clr that I've just build from the current master ~ few commits)

How long does it normally take to see a repro?

**Comment by stephentoub** [comment]
Seconds on my machine, using src from master this afternoon.

**Comment by jkotas** [comment]
Does you machine has AVX2? Also, are you using checked or release coreclr build?

**Comment by VSadov** [comment]
Using all release, yes machine has AVX2. I will try older build.

**Comment by jkotas** [comment]
Even simpler repro:
```
using System;
using System.Threading;

class Program
{
    static void GCLoop()
    {
        for (;;)
        {
            GC.Collect();
            Thread.Sleep(1);
        }
    }

    static void Main()
    {
        new Thread(GCLoop).Start();

        Span<byte> buffer1 = new byte[1024 * 1024];
        for (int i = 0; i < buffer1.Length; i++) buffer1[i] = (byte)i;
        Span<byte> buffer2 = new byte[buffer1.Length];
        buffer1.CopyTo(buffer2);

        while (true)
        {
           if (!buffer1.SequenceEqual(buffer2)) // should never be false...
           {
               Console.WriteLine("What in the world just happened?!");
            }
        }
    }
}
```

**Comment by VSadov** [comment]
The way I am trying to repro:
- created an app via  "dotnet new"
- pasted the repro code
- dotnet publish -c Release -r win-x64 --self-contained
- copied the fresh built release coreclr binaries into the app publish folder
- run

Is that the way you repro this?

I tried fresh, yesterdays, today mid-day bits. Tried all the repros above.
Verified in the debugger that I do run the repro code with the right runtime (in case there is some mismatch while copying)
And it does take the AVX2 path in SequenceEqual

**Comment by VSadov** [comment]
Machine (standard-issue Intel, you likely have similar) :

	Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz

	Base speed:	3.19 GHz
	Sockets:	1
	Cores:	6
	Logical processors:	12
	Virtualization:	Disabled
	Hyper-V support:	Yes
	L1 cache:	384 KB
	L2 cache:	1.5 MB
	L3 cache:	12.0 MB

**Comment by jkotas** [comment]
Looks like https://github.com/dotnet/runtime/issues/38974 finally caught up with us.

You may need to be on latest VS dogfood to see the crash and/or have processor without the ERMSB feature.

Here is what's happening:
1. Fully interruptible thread suspend happens to stop at this place in `SpanHelpers.SequenceEqual`:
```
00007ff9`c895aae7 c5fd10042e      vmovupd ymm0,ymmword ptr [rsi+rbp]
00007ff9`c895aaec c5fd100c2f      vmovupd ymm1,ymmword ptr [rdi+rbp] <--- Here
00007ff9`c895aaf1 c5fd74c1        vpcmpeqb ymm0,ymm0,ymm1
00007ff9`c895aaf5 c5fdd7c8        vpmovmskb ecx,ymm0
00007ff9`c895aaf9 83f9ff          cmp     ecx,0FFFFFFFFh
00007ff9`c895aafc 0f85e5000000    jne     System_Private_CoreLib!System.SpanHelpers.SequenceEqual(Byte ByRef, Byte ByRef, UIntPtr)+0xffffffff`a1614a47 (00007ff9`c895abe7)
```
2. The thread suspend helper happens to call memcpy at:
```
CoreCLR!memcpy+0x180 [d:\agent\_work\31\s\src\vctools\crt\vcruntime\src\string\amd64\memcpy.asm @ 300] 
CoreCLR!Thread::RedirectedHandledJITCase+0x27f [C:\runtime\src\coreclr\vm\threadsuspend.cpp @ 2708] 
CoreCLR!RedirectedHandledJITCaseForGCThreadControl_Stub+0x26 [C:\runtime\src\coreclr\vm\amd64\RedirectedHandledJITCase.asm @ 97] 
System_Private_CoreLib!System.SpanHelpers.SequenceEqual(Byte ByRef, Byte ByRef, UIntPtr)+0xffffffff`a161494c [C:\runtime\src\libraries\System.Private.CoreLib\src\System\SpanHelpers.Byte.cs @ 1676] 
```
The memcpy is optimized using AVX that modifies ymm0 register as side-effect:
```
00007ffa`28523a6c c5fe6f02        vmovdqu ymm0,ymmword ptr [rdx]
00007ffa`28523a70 c4a17e6f6c02e0  vmovdqu ymm5,ymmword ptr [rdx+r8-20h] ds:00000269`9ce6c9e0=60
00007ffa`28523a77 4981f800010000  cmp     r8,100h
00007ffa`28523a7e 0f86c4000000    jbe     CoreCLR!memcpy+0x258 (00007ffa`28523b48)
```
3. Nothing restores ymm0 to its original value when the thread resumes due to #38974. It makes SequenceEqual to return a bogus result.

**Comment by jkotas** [comment]
cc @mangod9

**Comment by VSadov** [comment]
I have ERMSB and I thought I am on the latest VS (it did not tell me that there is a new build, so I need to check).

**Comment by VSadov** [comment]
I did suspect that it is something that corrupts AVX registers. Suspension troubles or a GC hole would be unlikely to recover as in the very first repro and probably just lead to eventual AV instead.

I did not know about https://github.com/dotnet/runtime/issues/38974 , interesting.

Is it only about thread suspension code? GC could use AVX as well. In WKS mode GC runs "in-line" - without switching to a different thread, so I wonder if that would be another way to change `ymm`.

I am guessing the right solution might be just save/restore `CONTEXT_XSTATE`. Perhaps it is not a big expense next to the rest of the suspension stuff.

**Comment by jkotas** [comment]
> I am guessing the right solution might be just save/restore CONTEXT_XSTATE

Yep.

**Comment by jkotas** [comment]
>  In WKS mode GC runs "in-line" - without switching to a different thread

ymm registers is volatile register. When the GC runs in-line, it is triggered from allocation helper that is not expected to preserve volatile registers.

**Comment by geoffkizer** [comment]
Any idea why this just started showing up? I hadn't seen it until yesterday, now I'm seeing it all the time.

**Comment by VSadov** [comment]
> ymm registers is volatile register. When the GC runs in-line, it is triggered from allocation helper that is not expected to preserve volatile registers.

Right - that is why at least two threads were needed to repro. One to trigger, another to get suspended at inconvenient time and get corrupted.

**Comment by VSadov** [comment]
> Any idea why this just started showing up?

I am still unable to repro. Likely requires newest C runtime/compiler.

**Comment by jkotas** [comment]
> Any idea why this just started showing up?

I think it got triggered by memcpy changes in CRT. Are you on VS dogfood or have you installed VS update recently?

memcpy - shipping .NET 5.0 bits:
```
0:010> u coreclr!memcpy
coreclr!memcpy [D:\agent\_work\9\s\src\vctools\crt\vcruntime\src\string\amd64\memcpy.asm @ 129]:
00007fff`50a9d7b0 4c8bd9          mov     r11,rcx
00007fff`50a9d7b3 4c8bd2          mov     r10,rdx
00007fff`50a9d7b6 4983f810        cmp     r8,10h
00007fff`50a9d7ba 7654            jbe     coreclr!memcpy+0x60 (00007fff`50a9d810)
00007fff`50a9d7bc 4983f820        cmp     r8,20h
00007fff`50a9d7c0 762e            jbe     coreclr!memcpy+0x40 (00007fff`50a9d7f0)
00007fff`50a9d7c2 482bd1          sub     rdx,rcx
00007fff`50a9d7c5 730d            jae     coreclr!memcpy+0x24 (00007fff`50a9d7d4)
``` 

memcpy - master compiled using latest VS dogfood (Version 16.9.0 Preview 4.0 [30914.28.main])
```
0:011> u coreclr!memcpy
CoreCLR!memcpy [d:\agent\_work\31\s\src\vctools\crt\vcruntime\src\string\amd64\memcpy.asm @ 68]:
00007fff`455838f0 488bc1          mov     rax,rcx
00007fff`455838f3 4c8d1506c774ff  lea     r10,[CoreCLR!__acrt_rg_language_count (00007fff`44cd0000)]
00007fff`455838fa 4983f80f        cmp     r8,0Fh
00007fff`455838fe 0f870c010000    ja      CoreCLR!memcpy+0x120 (00007fff`45583a10)
00007fff`45583904 666666660f1f840000000000 nop word ptr [rax+rax]
00007fff`45583910 478b8c8210d2ca00 mov     r9d,dword ptr [r10+r8*4+0CAD210h]
00007fff`45583918 4d03ca          add     r9,r10
00007fff`4558391b 41ffe1          jmp     r9
...
uses AVX 
...
```

**Comment by geoffkizer** [comment]
> have you installed VS update recently?

Yeah, yesterday, just before this started happening. Makes sense, thanks.

**Comment by mangod9** [comment]
just catching up, so it appears this will be resolved by #38974 , and wouldn't need any other change?

**Comment by danmosemsft** [comment]
When we service existing releases, do we use a pinned CRT version? Or will they start to hit this too.

**Comment by hoyosjs** [comment]
> When we service existing releases, do we use a pinned CRT version? Or will they start to hit this too.

We don't pin a CRT.

**Comment by mangod9** [comment]
umm.. that is interesting. that would imply backporting to 3.1 or lower. Also wonder if the memcpy change now implies an AVX requirement, or if they are doing a dynamic check for it.

**Comment by jkotas** [comment]
They are doing dynamic check.

We should understand the VS update schedule on the build machines. If/once the build machine are updated to VS Version 16.9.0, this will become a blocking problem.

**Comment by hoyosjs** [comment]
Right now we have 16.8.3 and 16.9 P2 on the lab machines. We build against 16.8.3 as of Wednesday.

**Comment by VSadov** [comment]
Is anyone working on making a fix for this? (or #38974, which seems the same issue)
I'd be interested to take a look.

**Comment by jkotas** [comment]
I do not think anybody is working on it. It would great if you can pick it up. Thank you!

**Comment by mangod9** [comment]
@VSadov I had been working on a possible fix which I have just pushed here: https://github.com/mangod9/runtime/commits/gh38974. It isnt complete yet (need to figure out how to correctly delete the allocated Context) , so please take a look if thats viable. You are probably in a better position to fix than me :)

**Comment by VSadov** [comment]
@mangod9 Thanks! I will try using your changes as a starting point. It looks like a plausible direction, although I did not get into every detail yet.

**Comment by mangod9** [comment]
yup sounds good. I have validated that the fix works for the simplified repro from Jan above, just need to ensure that the allocation is ok at that spot and cause no leaks. I have left comments on the commit as such. Thx.

**Comment by benaadams** [comment]
Will this also effect the `SequenceEquals` in `System.Memory` on .NET Framework? i.e.. will that also be affected by the memcpy changes in CRT?

**Comment by jkotas** [comment]
This potentially affects any code that uses AVX, e.g. any code that uses `Vector<T>` is potentially affected. This will affect .NET Framework once/if the C++ compiler used to build .NET Framework is updated. We will keep that in mind.

**Comment by VSadov** [comment]
Fixed in https://github.com/dotnet/runtime/pull/47212

**Comment by danmosemsft** [comment]
Do we need a new issue, or to keep this open, for any servicing?

**Comment by mangod9** [comment]
New issue would be better for servicing. I will create one for 5.0, and can keep that around for more down-level ports if required

**Comment by VSadov** [comment]
perhaps better to open a separate issue and track all the releases where this needs to be ported.

**Comment by mangod9** [comment]
Actually we can repurpose https://github.com/dotnet/runtime/issues/38974, for that. I will reopen and mark for 5.0.x.

**Comment by VSadov** [comment]
I think any CLR version that supports AVX intrinsics and still receives servicing fixes may need to get this fix at some point.

**Comment by mangod9** [comment]
@jeffschwMSFT as fyi.

**Comment by jarodrig** [comment]
Just wanted to say:  CRT memcpy was updated to use AVX registers -- this change shipped in VS 2019 P3.
